### PR TITLE
include: zephyr: drivers: i3c: Add write_buffer_*_cb target callbacks

### PR DESCRIPTION
**Pull request by myguitar**
Some controllers receive data from a Rx FIFO, which requires a minimum packet size (e.g., 4-bytes dword). In this case, byte-by-byte callbacks may not be suitable for FIFO-based received data. To address this issue, two new callbacks are introduced:
- write_buffer_requested_cb: allows an application to allocate a received buffer
- write_buffer_completed_cb: notifies the application when the received data is complete

### CODE REVIEW

**Review by XenuIsWatching — CHANGES_REQUESTED** [review]
I find it very hard to believe that a IP would force a minimum packet size even with a 32b fifo size. What IP are you referring to here? I also do not understand how a "controller" can request a buffer. How do you expect 'requested size' to be known before hand and for it to be passed back to the callback?

I also do not see any use of these Driver APIs with this PR. I'm not sure what the "official policy is here", but I'm not a fan of adding Driver APIs unless if there is a in-tree driver using it. Do you have another driver that is in-progress?

It also may need to be considered in the case the controller does an "abort" in the middle of a private read for the completed case as it would be valuable to know.... but the general i3c target API is probably missing that definition as well for the byte by byte case... 😢

**Comment by dcpleung** [comment]
Shouldn't these be handled by the driver itself? Why do you need to expose these?

**Comment by myguitar** [comment]
> I find it very hard to believe that a IP would force a minimum packet size even with a 32b fifo size. What IP are you referring to here? I also do not understand how a "controller" can request a buffer. How do you expect 'requested size' to be known before hand and for it to be passed back to the callback?
> 
> I also do not see any use of these Driver APIs with this PR. I'm not sure what the "official policy is here", but I'm not a fan of adding Driver APIs unless if there is a in-tree driver using it. Do you have another driver that is in-progress?
> 
> It also may need to be considered in the case the controller does an "abort" in the middle of a private read for the completed case as it would be valuable to know.... but the general i3c target API is probably missing that definition as well for the byte by byte case... 😢



> I find it very hard to believe that a IP would force a minimum packet size even with a 32b fifo size. What IP are you referring to here? I also do not understand how a "controller" can request a buffer. How do you expect 'requested size' to be known before hand and for it to be passed back to the callback?
> 
> I also do not see any use of these Driver APIs with this PR. I'm not sure what the "official policy is here", but I'm not a fan of adding Driver APIs unless if there is a in-tree driver using it. Do you have another driver that is in-progress?
> 
> It also may need to be considered in the case the controller does an "abort" in the middle of a private read for the completed case as it would be valuable to know.... but the general i3c target API is probably missing that definition as well for the byte by byte case... 😢

Let me take a step back about `minimum packet size`. It's FIFO entry size actually. I am working on synopsis i3c controller and planning to upstream it to the Zephyr main. Sorry, I should have sent a pull request with the driver. I will do it soon. The synopsis i3c controller has dword FIFO entry with multiple entries, and it triggers an interrupt when the Rx FIFO data is ready with available data size. So, the `requested size` is known at the time of interrupt. It can use the byte by byte callback, but it requires additional copy from a temporary buffer to the actual target buffer.
The synopsis controller sends an error status interrupt in case of `abort`, but that's good idea to have an another callback for the error cases. I will send out driver code along with this. Please let me know if you have any suggestions or concerns. Thanks.

**Comment by myguitar** [comment]
> Shouldn't these be handled by the driver itself? Why do you need to expose these?

It's possible in the driver itself, but a common interface for the FIFO based controller would be easy to maintain.

**Comment by XenuIsWatching** [comment]
> Let me take a step back about `minimum packet size`. It's FIFO entry size actually. I am working on synopsis i3c controller and planning to upstream it to the Zephyr main. Sorry, I should have sent a pull request with the driver. I will do it soon. The synopsis i3c controller has dword FIFO entry with multiple entries, and it triggers an interrupt when the Rx FIFO data is ready with available data size. So, the `requested size` is known at the time of interrupt. It can use the byte by byte callback, but it requires additional copy from a temporary buffer to the actual target buffer. The synopsis controller sends an error status interrupt in case of `abort`, but that's good idea to have an another callback for the error cases. I will send out driver code along with this. Please let me know if you have any suggestions or concerns. Thanks.

I see, I'm very familiar with that IP and was working with it for a while. I think you're with Google, right? I remember before I started working with it. I remember some people I chat with there at the EOSS 23 conference in Prague had mentioned they had a driver internally that supported target mode, but then I found out a few months later that I was then start working with it at my company. I pinged a few people at Google over Discord about what their upstream plans were for that synopsys I3C, but none knew anything or didn't even know they had a driver for that. I had a deadline I had to work through... so I just started from mostly scratch with writing my own driver for that synopsys I3C. I did get a driver written for it in a couple of weeks, but we had a issue had to change to another vendor for I3C and I had to abandoned development of that driver although it was feature complete and I even had a draft diff PR on GitHub... but I never got around to testing it.... months later I now see Intel publish their synopsys driver a GitHub PR (although it had quite a few issues with it as it appears it was originally written before the I3C API was more flushed out as it is now currently). Now I'm hearing Google wants to upstream their driver... which does give a bit of frustration here as that would now be the 4th PR for the same I3C along with Samsung (https://github.com/zephyrproject-rtos/zephyr/pull/50847), Meta (https://github.com/zephyrproject-rtos/zephyr/pull/64528), Intel (https://github.com/zephyrproject-rtos/zephyr/pull/66519), and Google all trying to publish their drivers for the same thing, and I also heard privately that a 5th Company is also looking at developing a driver for the same IP.

For me... that synopsys I3C driver is being resurrected back up this month from the coding graveyard to go through testing this month. I did take a look already at the Samsung and Intel driver and it appears the one in the https://github.com/zephyrproject-rtos/zephyr/pull/64528 has more the features and the API implemented in the PR that I wrote long ago (althrough there is a few minor things that need to be incorporated from the Intel driver in to that PR)
 It's probably best to just have 1 PR that everyone comments for issues that are found.... although I know Google has told me that they got 'Target Mode' working...and I know that is a feature that I do not expect to test as rigorously as for controller mode operation. I'm not sure what the timeline would be for testing for us and upstreaming for ya'll.... but we need to make sure that we are all getting what we'd like out of this driver, and the more eyes/developers we have looking at one source the better driver we will all have :) 
For that PR, I do have quite a few TODOs in for that Target Mode, we may want to work together to make sure it does what you're needs are (and to make sure it actually works 😄 ) . It could be in both of our advantages to merge your Target Code in when it is ready.
As for the timeline of that PR, i do expect testing go over the next few weeks and for to be brought out of the "draft" stage immediately after that. If you want and have time, you can start looking at the https://github.com/zephyrproject-rtos/zephyr/pull/64528 PR for comments and concerns. I also looked through and compared https://github.com/zephyrproject-rtos/zephyr/pull/66519 to make sure that https://github.com/zephyrproject-rtos/zephyr/pull/64528 contained all the features that had.

**Comment by myguitar** [comment]
> > Let me take a step back about `minimum packet size`. It's FIFO entry size actually. I am working on synopsis i3c controller and planning to upstream it to the Zephyr main. Sorry, I should have sent a pull request with the driver. I will do it soon. The synopsis i3c controller has dword FIFO entry with multiple entries, and it triggers an interrupt when the Rx FIFO data is ready with available data size. So, the `requested size` is known at the time of interrupt. It can use the byte by byte callback, but it requires additional copy from a temporary buffer to the actual target buffer. The synopsis controller sends an error status interrupt in case of `abort`, but that's good idea to have an another callback for the error cases. I will send out driver code along with this. Please let me know if you have any suggestions or concerns. Thanks.
> 
> I see, I'm very familiar with that IP and was working with it for a while. I think you're with Google, right? I remember before I started working with it. I remember some people I chat with there at the EOSS 23 conference in Prague had mentioned they had a driver internally that supported target mode, but then I found out a few months later that I was to then start working with it at my company. I pinged a few people at Google over Discord about what their upstream plans were for that synopsys I3C, but none knew anything or didn't even know they had a driver for that. I had a deadline I had to work through... so I just started from mostly scratch with writing my own driver for that synopsys I3C. I did get a driver written for it in a couple of weeks, but we had a issue had to change to another vendor for I3C and I had to abandoned development of that driver although it was feature complete and I even had a draft diff PR on GitHub... but I never got around to testing it.... months later I now see Intel publish their synopsys driver a GitHub PR (although it had quite a few issues with it as it appears it was originally written before the I3C API was more flushed out as it is now currently). Now I'm hearing Google wants to upstream their driver... which does give a bit of frustration here as that would now be the 4th PR for the same I3C along with Samsung (#50847), Meta (#64528), Intel (#66519), and Google all trying to independently publish their drivers for the same thing, and I also heard privately that a 5th Company is also currently looking at developing a driver for the same IP.
> 
> For me... that synopsys I3C driver is being resurrected back up this month from the coding graveyard to go through testing this month. I did take a look already at the Samsung and Intel driver and it appears the one in the #64528 has more the features and the API implemented in the PR that I wrote long ago (althrough there is a few minor things that need to be incorporated from the Intel driver in to that PR) It's probably best to just have 1 PR that everyone comments for issues that are found.... although I know Google has told me that they got 'Target Mode' working...and I know that is a feature that I do not expect to test as rigorously as for controller mode operation. I'm not sure what the timeline would be for testing for us and upstreaming for ya'll.... but we need to make sure that we are all getting what we'd like out of this driver, and the more eyes/developers we have looking at one source the better driver we will all have :) For that PR, I do have quite a few TODOs in for that Target Mode, we may want to work together to make sure it does what you're needs are (and to make sure it actually works 😄 ) . It could be in both of our advantages to merge your Target Code in when it is ready. As for the timeline of that PR, i do expect testing go over the next few weeks and for to be brought out of the "draft" stage immediately after that. If you want and have time, you can start looking at the #64528 PR for comments and concerns. I also looked through and compared #66519 to make sure that #64528 contained all the features that had.

I totally understand your concern and was also in the same boat in 2022. Waiting for synopsys I3C zephyr driver merged at the time, but there was no good news I have been waiting for so started developing the I3C driver. Now Google has the synopsys I3C driver supporting both `Controller Mode` and `Target Mode` based on zephyr 3.6.0.
I am happy to collaborate with you and zephyr contributors for the synopsys I3C driver. Sure, let me wait for your `Controller Mode` driver and I will start looking at the #64528 PR. Once it's ready, I will merge `Target Mode` driver based on yours. Appreciate sharing your concerns and plans. Please feel free to close this PR. Thanks.

**Pull request closed without merging**
